File: app/src/main/java/com/hy/android/fragment/HomeFragment.java
package com.hy.android.fragment;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.LinearLayout;

import com.chad.library.adapter.base.BaseQuickAdapter;
import com.hy.android.R;
import com.hy.android.activity.ArticleDetailActivity;
import com.hy.android.adapter.HomeAdapter;
import com.hy.android.base.BaseFragment;
import com.hy.android.bean.BannerData;
import com.hy.android.bean.BaseResponse;
import com.hy.android.bean.HomeData;
import com.hy.android.bean.HomeDataList;
import com.hy.android.net.BaseObserver;
import com.hy.android.net.RetrofitHelper;
import com.hy.android.utils.Constants;
import com.hy.android.utils.GlideImageLoader;
import com.youth.banner.Banner;
import com.youth.banner.BannerConfig;
import com.youth.banner.Transformer;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by Administrator on 2018/4/17.
 */

public class HomeFragment extends BaseFragment {

    private static final String TAG = "HomeFragment";
    private SwipeRefreshLayout swipeRefreshLayout;
    private RecyclerView mRecyclerView;
    private HomeAdapter mAdapter;
    private List<HomeData> homeDatas;
    private List<BannerData> bannerDatas;
    private Banner mBanner;

    @Override
    public int getContentLayout() {
        return R.layout.fragment_home;
    }

    @Override
    public void initData() {

    }

    @Override
    public void bindView(View view, Bundle savedInstanceState) {
        homeDatas = new ArrayList<>();
        bannerDatas = new ArrayList<>();

        swipeRefreshLayout = view.findViewById(R.id.swipeRefreshLayout);
        //banner
        LinearLayout mHeaderGroup = ((LinearLayout) LayoutInflater.from(getActivity()).inflate(R.layout.home_banner, null));
        mBanner = mHeaderGroup.findViewById(R.id.head_banner);
        mHeaderGroup.removeView(mBanner);
        //list
        mRecyclerView = view.findViewById(R.id.recyclerView);
        mAdapter = new HomeAdapter(R.layout.home_list_item, homeDatas);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        mAdapter.addHeaderView(mBanner);
        mRecyclerView.setAdapter(mAdapter);
        mAdapter.setOnLoadMoreListener(new BaseQuickAdapter.RequestLoadMoreListener() {
            @Override
            public void onLoadMoreRequested() {
                int page = mAdapter.getData().size() / 20 + 1;
                getHomeList(page);
            }
        }, mRecyclerView);

        swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                refreshData();
            }
        });
        getBanner();
        getHomeList(0);

        mAdapter.setOnItemClickListener(new BaseQuickAdapter.OnItemClickListener() {
            @Override
            public void onItemClick(BaseQuickAdapter adapter, View view, int position) {
                ArticleDetailActivity.startActivity(getActivity(), homeDatas.get(position).title, homeDatas.get(position).link);
            }
        });
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

    }

    private void refreshData() {
        swipeRefreshLayout.setRefreshing(true);
        mAdapter.setEnableLoadMore(false);
        getHomeList(0);
    }

    private void getBanner() {
        RetrofitHelper.getInstance().getApiService(Constants.ANDROID_URL).getBannerData()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new BaseObserver<BaseResponse<List<BannerData>>>() {
                    @Override
                    public void onSuccess(BaseResponse<List<BannerData>> response) {
                        if (response.getErrorCode() == 0 && response.getData().size() > 0) {
                            bannerDatas.clear();
                            bannerDatas.addAll(response.getData());
                            showBannerData(bannerDatas);
                        } else {
                            toast(response.getErrorMsg());
                        }
                    }

                    @Override
                    public void onFail(Throwable e) {
                        Log.e(TAG, e.getMessage());
                    }
                });
    }

    private void getHomeList(final int page) {
        RetrofitHelper.getInstance().getApiService(Constants.ANDROID_URL).getHomeList(page)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new BaseObserver<BaseResponse<HomeDataList>>() {
                    @Override
                    public void onSuccess(BaseResponse<HomeDataList> response) {
                        if (response.getErrorCode() == 0 && response.getData() != null) {

                            int total = response.getData().total;

                            if (total == 0) {
                                return;
                            }

                            if (total < response.getData().size) {
                                mAdapter.replaceData(response.getData().datas);
                                mAdapter.loadMoreComplete();
                                mAdapter.loadMoreEnd();
                                mAdapter.setEnableLoadMore(false);
                                return;
                            }

                            Log.e("data==", total + "--" + response.getData().size);

                            if (response.getData().offset >= total || response.getData().size >= total) {
                                mAdapter.loadMoreEnd();
                                return;
                            }
                            if (swipeRefreshLayout.isRefreshing()) {
                                mAdapter.replaceData(response.getData().datas);
                            } else {
                                mAdapter.addData(response.getData().datas);
                            }
                            mAdapter.loadMoreComplete();
                            mAdapter.setEnableLoadMore(true);
                            mAdapter.notifyDataSetChanged();
                        } else {
                            toast(response.getErrorMsg());
                        }
                        swipeRefreshLayout.setRefreshing(false);
                    }

                    @Override
                    public void onFail(Throwable e) {
                        swipeRefreshLayout.setRefreshing(false);
                        mAdapter.setEnableLoadMore(false);
                        mAdapter.loadMoreFail();
                        Log.e(TAG, e.getMessage());
                    }
                });
    }

    public void showBannerData(List<BannerData> bannerDataList) {

        List<String> bannerImageList = new ArrayList<>();
        List<String> mBannerTitleList = new ArrayList<>();
        for (BannerData data : bannerDataList) {
            bannerImageList.add(data.imagePath);
            mBannerTitleList.add(data.title);
        }
        //??????banner??????
        mBanner.setBannerStyle(BannerConfig.NUM_INDICATOR_TITLE);
        //?????????????????????
        mBanner.setImageLoader(new GlideImageLoader());
        //??????????????????
        mBanner.setImages(bannerImageList);
        //??????banner????????????
        mBanner.setBannerAnimation(Transformer.DepthPage);
        //????????????????????????banner???????????????title??????
        mBanner.setBannerTitles(mBannerTitleList);
        //??????????????????????????????true
        mBanner.isAutoPlay(true);
        //??????????????????
        mBanner.setDelayTime(bannerDataList.size() * 400);
        //???????????????????????????banner???????????????????????????
        mBanner.setIndicatorGravity(BannerConfig.CENTER);

        //banner?????????????????????????????????????????????
        mBanner.start();
    }

    @Override
    public void onResume() {
        super.onResume();
        if (mBanner != null) {
            mBanner.startAutoPlay();
        }
    }

    @Override
    public void onStop() {
        super.onStop();
        if (mBanner != null) {
            mBanner.stopAutoPlay();
        }
    }
}

File: WanAndroid/src/main/java/com/kt/android/ui/home/HomeFragment.kt
package com.kt.android.ui.home

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.android.common.base.BaseFragment
import com.android.common.ui.WebViewActivity
import com.kt.android.R
import com.kt.android.adapter.HomeAdapter
import com.kt.android.bean.BannerData
import com.kt.android.bean.HomeData
import com.kt.android.databinding.FragmentHomeBinding
import com.kt.android.util.GlideImageLoader
import com.kt.android.util.SpacesItemDecoration
import com.kt.android.util.dp
import com.youth.banner.Banner
import com.youth.banner.BannerConfig
import com.youth.banner.Transformer
import java.util.*

class HomeFragment : BaseFragment<FragmentHomeBinding>() {
    companion object {
        private const val TAG = "HomeFragment"
    }
    private val pageSize = 20
    private lateinit var homeViewModel: HomeViewModel
    private var mBanner: Banner? = null
    private var mAdapter: HomeAdapter? = null
    private var homeDataList: ArrayList<HomeData>? = null
    private var bannerDataList: ArrayList<BannerData>? = null

    override fun getLayoutID(): Int = R.layout.fragment_home

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        homeViewModel = ViewModelProvider(this).get(HomeViewModel::class.java)
        initData()
        initView()
    }


    private fun initData() {
        homeViewModel.requestBannerData()
        homeViewModel.requestHomeList(0)
    }

    private fun initView() {
        homeDataList = ArrayList<HomeData>()
        bannerDataList = ArrayList<BannerData>()
        val mHeaderGroup =
            LayoutInflater.from(activity).inflate(R.layout.home_banner, null) as LinearLayout
        mBanner = mHeaderGroup.findViewById(R.id.head_banner)
        mHeaderGroup.removeView(mBanner)

        mBind.recyclerView.layoutManager = LinearLayoutManager(activity)
        mBind.recyclerView.addItemDecoration(SpacesItemDecoration(8f.dp.toInt()))
        mAdapter = HomeAdapter(R.layout.home_list_item, homeDataList)
        mAdapter!!.addHeaderView(mBanner!!)
        mBind.recyclerView.adapter = mAdapter

        mAdapter?.setOnItemClickListener { _, _, position ->
            WebViewActivity.startActivity(activity,mAdapter?.getItem(position)?.link)
        }

        mBind.swipeRefreshLayout.setOnRefreshListener {
            mBind.swipeRefreshLayout.isRefreshing = true
            mAdapter?.loadMoreModule?.isEnableLoadMore = false
            homeViewModel.requestHomeList(0)
        }

        mAdapter?.loadMoreModule?.setOnLoadMoreListener {
            mAdapter?.let {
                mBind.swipeRefreshLayout.isRefreshing = false
                it.loadMoreModule.isEnableLoadMore = true
                val page = it.data.size / pageSize + 1
                homeViewModel.requestHomeList(page)
            }
        }
        mAdapter?.loadMoreModule?.isAutoLoadMore = true
        mAdapter?.loadMoreModule?.isEnableLoadMoreIfNotFullPage = false

        homeViewModel.bannerLiveData.observe(viewLifecycleOwner) { result ->
            val bannerResult = result.getOrNull()
            if (null == bannerResult || bannerResult.data.isEmpty()) {
                Log.d(TAG, "get banner data failed")
                return@observe
            }
            bannerDataList?.let {
                it.addAll(bannerResult.data)
                showBannerData(it)
            }
        }

        homeViewModel.homeLiveData.observe(viewLifecycleOwner) { result ->
            val bannerResult = result.getOrNull()
            if (null == bannerResult || bannerResult.data.datas.isEmpty()) {
                Log.d(TAG, "get home data failed")
                return@observe
            }
            mBind.swipeRefreshLayout.isRefreshing = false
            mAdapter?.loadMoreModule?.isEnableLoadMore = true
            val data = bannerResult.data.datas
            Log.d(TAG, "data size -- ${data.size}")
            val page = data.size / pageSize + 1
            if (page == 0) {
                mAdapter?.setList(data)
            } else {
                mAdapter?.addData(data)
            }
            if (data.size < pageSize) {
                mAdapter?.loadMoreModule?.loadMoreEnd()
            } else {
                mAdapter?.loadMoreModule?.loadMoreComplete()
            }
        }
    }

    private fun showBannerData(bannerDataList: List<BannerData>) {
        val bannerImageList: MutableList<String?> = ArrayList()
        val mBannerTitleList: MutableList<String> = ArrayList()
        for (data in bannerDataList) {
            bannerImageList.add(data.imagePath)
            mBannerTitleList.add(data.title)
        }
        mBanner?.let {
            //设置banner样式
            it.setBannerStyle(BannerConfig.NUM_INDICATOR_TITLE)
            //设置图片加载器
            it.setImageLoader(GlideImageLoader())
            //设置图片集合
            it.setImages(bannerImageList)
            //设置banner动画效果
            it.setBannerAnimation(Transformer.DepthPage)
            //设置标题集合（当banner样式有显示title时）
            it.setBannerTitles(mBannerTitleList)
            //设置自动轮播，默认为true
            it.isAutoPlay(true)
            //设置轮播时间
            it.setDelayTime(3000)
            //设置指示器位置（当banner模式中有指示器时）
            it.setIndicatorGravity(BannerConfig.CENTER)
            //banner设置方法全部调用完毕时最后调用
            it.start()
        }
    }
}